File: src/main/scala/bank/Bank.scala
package bank

import akka.actor.SupervisorStrategy.{Restart, Resume}
import akka.actor._
import bank.Bank._

import scala.concurrent.duration._

/**
 * Created by Rahul Shukla <rahul.shukla@synerzip.com> on 8/10/15.
 */

object Bank {

  sealed trait BankMessage

  case class Balance(number: String) extends BankMessage

  case class OpenAccount(number: String) extends BankMessage

  case class Withdraw(number: String, amount: Int) extends BankMessage

  case class Deposit(number: String, amount: Int) extends BankMessage

  case class Transfer(from: String, to: String, amount: Int) extends BankMessage

  def props(name: String): Props = {
    return Props(Bank(name))
  }
}

case class Bank(name: String) extends AbstractLoggingActor {

  def createAccount(number: String) = {
    log.info(s"Creating Account with number $number")
    context.actorOf(Account.props(number), name = number)
  }

  def withdraw(number: String, amount: Int) = {
    context.actorSelection(number) ! Account.Withdraw(amount)
  }

  def deposit(number: String, amount: Int) = {
    context.actorSelection(number) ! Account.Deposit(amount)
  }

  def balance(number: String) = {
    context.actorSelection(number) ! Account.Balance
  }

  def transfer(from: String, to: String, amount: Int) = {
    log.info(s"Transfering $amount from $from to $to")
    context.actorSelection(from) ! Account.Transfer( to ,amount)
  }

  override def receive = {
    case OpenAccount(number) => createAccount(number)
    case Withdraw(number, amount) => withdraw(number, amount)
    case Deposit(number, amount) => deposit(number, amount)
    case Transfer(from, to, amount) => transfer(from, to, amount)
    case Balance(number) => balance(number)
  }

  override def supervisorStrategy = OneForOneStrategy(maxNrOfRetries = 10, withinTimeRange = 1 minute) {
    case _: Account.OverdrawException => Resume
    case _: Exception => Restart
  }

}

File: src/main/java/bank1/Bank.java
package bank1;

import akka.actor.AbstractLoggingActor;
import akka.actor.Props;
import akka.japi.pf.ReceiveBuilder;
import scala.PartialFunction;
import scala.runtime.BoxedUnit;

/**
 * Created by Rahul Shukla <rahul.shukla@synerzip.com> on 9/10/15.
 */
public class Bank extends AbstractLoggingActor {

    private final String name;

    public Bank(String name) {

        this.name = name;
    }

    static interface BankMessage {
    }

    static class Balance implements BankMessage {
        public final String number;

        public Balance(String number) {
            this.number = number;
        }
    }

    static class OpenAccount implements BankMessage {
        public final String number;

        public OpenAccount(String number) {
            this.number = number;
        }
    }

    static class Withdraw implements BankMessage {
        public final String number;
        public final int amount;

        public Withdraw(String number, int amount) {
            this.number = number;
            this.amount = amount;
        }
    }

    static class Deposit implements BankMessage {
        public final String number;
        public final int amount;

        public Deposit(String number, int amount) {
            this.number = number;
            this.amount = amount;
        }
    }

    static class Transfer implements BankMessage

    {
        public final String from;
        public final String to;
        public final int amount;

        public Transfer(String from, String to, int amount) {
            this.from = from;
            this.to = to;
            this.amount = amount;
        }
    }

    public static Props props(String name) {
        return Props.create(Bank.class, () -> new Bank(name));
    }

    @Override
    public PartialFunction<Object, BoxedUnit> receive() {
        return ReceiveBuilder
                .match(Balance.class, b -> balance(b.number))
                .match(Deposit.class, d -> deposit(d.number, d.amount))
                .match(Withdraw.class, w -> withdraw(w.number, w.amount))
                .match(Transfer.class, t -> transfer(t.from, t.to, t.amount))
                .match(OpenAccount.class, a -> create(a.number))
                .build();
    }

    private void create(String number) {
        getContext().actorOf(Account.props(number), number);
    }

    private void balance(String number) {
        getContext().actorSelection(number).tell(new Account.Balance(), self());
    }

    private void transfer(String from, String to, int amount) {
        getContext().actorSelection(from).tell(new Account.Transfer(to, amount), self());
    }

    private void withdraw(String number, int amount) {
        getContext().actorSelection(number).tell(new Account.Withdraw(amount), self());
    }

    private void deposit(String number, int amount) {
        getContext().actorSelection(number)
                .tell(new Account.Deposit(amount), self());
    }
}
